server/index-node/src/service.rs

```diff
@@ -118,46 +123,47 @@ where
                             max_depth: 100,
                             max_first: std::u32::MAX,
                         },
-                    )))
+                    ))
                 })
-                .then(move |result| {
-                    let elapsed = start.elapsed().as_millis();
-                    match result {
-                        Ok(_) => info!(
-                            result_logger,
-                            "GraphQL query served";
-                            "query_time_ms" => elapsed,
-                            "code" => LogCode::GraphQlQuerySuccess,
-                        ),
-                        Err(ref e) => error!(
-                            result_logger,
-                            "GraphQL query failed";
-                            "error" => e.to_string(),
-                            "query_time_ms" => elapsed,
-                            "code" => LogCode::GraphQlQueryFailure,
-                        ),
-                    }
-                    IndexNodeResponse::new(result)
-                }),
-        )
+            })
+            .then(move |result| {
+                let elapsed = start.elapsed().as_millis();
+                match result {
+                    Ok(_) => info!(
+                        result_logger,
+                        "GraphQL query served";
+                        "query_time_ms" => elapsed,
+                        "code" => LogCode::GraphQlQuerySuccess,
+                    ),
+                    Err(ref e) => error!(
+                        result_logger,
+                        "GraphQL query failed";
+                        "error" => e.to_string(),
+                        "query_time_ms" => elapsed,
+                        "code" => LogCode::GraphQlQueryFailure,
+                    ),
+                }
+                IndexNodeResponse::new(result).compat()
+            })
+            .boxed()
     }
 
     // Handles OPTIONS requests
     fn handle_graphql_options(&self, _request: Request<Body>) -> IndexNodeServiceResponse {
-        Box::new(future::ok(
-            Response::builder()
+        Box::pin(async {
```

Comment:
There's an explanation [here](https://rust-lang.github.io/async-book/04_pinning/01_chapter.html).

The thing is so far all types in Rust were safe to move, but `async` blocks are a new language feature called generators, which would not be safe to poll again after being moved when they hold references to themselves. So to make sure that futures are not going to move between calls to poll, `Future::poll` takes `self: Pin<&mut Self>`. Most types we don't have to pin ourselves because they are `Unpin` which means they don't care if they are pinned or not, but generators need to be pinned with `Box::pin`. Yay, I got it, I think.

Reply:
Haha, still have to wrap my head around it. Thanks for the explanation!